src/toniarts/openkeeper/animation/PoseTrack.java

```diff
@@ -57,9 +52,9 @@
 /**
  * A single track of pose animation associated with a certain mesh.
  */
-public final class PoseTrack implements Track, ClonableTrack {
+public final class PoseTrack extends MorphTrack {

```

Comment:
I guess this could be a first step towards solving https://github.com/tonihele/OpenKeeper/pull/409#discussion_r918624619.
AFAICT you'd need to define `MorphTarget`s for the target Geometry instead of custom `Pose`s?

Reply:
The jME morph targets work a bit differently. If I remember correctly, they are hardware accelerated. But there was some reason that they don't really work with our way of thinking... Was it that we would then have millions of morph targets per model? And it just didn't work out... Something like this? And the hardware acceleration worked only up until 7 morph targets...

Reply:
Yeah I think it's similar to what I do in blender, 1 morph target per frame and weights like (0,0,0,1,0,0...), so quite a few.
But maybe MorphTrack could otherwise be abused into what we need.

https://github.com/jMonkeyEngine/jmonkeyengine/blob/af375391ea385a061444ca03262e1a4720c3274d/jme3-core/src/main/java/com/jme3/anim/tween/action/ClipAction.java#L49-L53

Reply:
Thanks!

The problem is that the whole PoseTrack thing is even harder to understand than KMF itself. All those previous and next frame refs in FrameInfo and Pose, all the maps and copy paste calls in:
https://github.com/tonihele/OpenKeeper/blob/604a33a377c6c069c7cf60dfea3bdaa7711b9899/src/toniarts/openkeeper/tools/convert/KmfModelLoader.java#L414
There also seems to be quite some redundancy:
https://gist.github.com/Trass3r/cdd614ec24c48f94969389492117fa0d#file-piranha_swim-xml-L163

Reply:
Looks like we were already passing through this code, it works cause in `getDataAtTime` we never set the output weights parameter and then `setMorphState` is a no-op:
```java
    private void interpolateMorphTrack(double t, MorphTrack track) {
        Geometry target = track.getTarget();
        float[] weights = target.getMorphState();
        track.getDataAtTime(t, weights);
        target.setMorphState(weights);
```

Btw the shader is indeed limited to a handful of morph targets to blend but you don't really need that many, and the system also falls back to CPU morph target merging as needed:
https://github.com/jMonkeyEngine/jmonkeyengine/blob/af375391ea385a061444ca03262e1a4720c3274d/jme3-core/src/main/java/com/jme3/anim/MorphControl.java#L157-L158